### ISSUE TITLE
HSL units are being changed, resulting in invalid colors

### ISSUE THREAD

**Issue by Enteleform** [vivliostyle/vivliostyle.js#1012]
## Describe the bug
When using CSS variables which have been assigned HSL values, the percentage units are changed to viewport width units, which results in invalid colors.

## Expected behavior
HSL values should be preserved with valid definitions.

## To Reproduce

[github.com/enteleform-codesandbox/0018--Vivliostyle--HSL](https://github.com/enteleform-codesandbox/0018--Vivliostyle--HSL)

The linked reproduction renders the DOM content as expected, but the CSS variable issue is present:

![repro-hsl](https://user-images.githubusercontent.com/10906415/193488020-e5cca7a7-a9bd-4ad2-9dce-fe281180f7a2.png)

In the repro, it seems that the color values are being rendered to RGB instead of maintaining the variable references:

![repro-rgb](https://user-images.githubusercontent.com/10906415/193489557-a7c1edf1-ac76-4f7f-a4da-6583e93424d2.png)

However, in the actual project where I ran into this issue, the variables are not being rendered to RGB

![project-css](https://user-images.githubusercontent.com/10906415/193489513-b84cf58f-3b66-49d1-b7c5-d3652b51aff5.png)

so the values are being resolved to the invalid HSL definitions, which results in invisible DOM elements.

## Desktop
 - OS: Windows 10
 - Browser: Chrome
 - Version: 106.0.5249.91

**Comment by MurakamiShinyu** [comment]
@Enteleform Thanks for finding the bug!

Fixed the bug in #1013.

Tested with your example:
https://vivliostyle-git-fix-issue1012-vivliostyle.vercel.app/#src=https://raw.githack.com/enteleform-codesandbox/0018--Vivliostyle--HSL/main/index.html

![Screen Shot 2022-10-03 at 23 42 30](https://user-images.githubusercontent.com/3324737/193606418-c839d459-14fc-43ae-a6f4-409a8adb8276.png)

Additional tests:

```html
<style>
:root {
  color: rgb(0%, 50%, 0%);
}
</style>
<p>This text should be green.</p>
```

- Before this fix (v2.18.3): https://vivliostyle.github.io/viewer/v2.18.3/#src=data:,%3Cstyle%3E%20:root%20%7B%20%20%20color:%20rgb(0%25,%2050%25,%200%25);%20%7D%20%3C/style%3E%20%3Cp%3EThis%20text%20should%20be%20green.%3C/p%3E
  ![Screen Shot 2022-10-03 at 23 56 23](https://user-images.githubusercontent.com/3324737/193609613-263ce178-0fde-4f01-9375-ff46ab149658.png)
- After this fix:  https://vivliostyle-git-fix-issue1012-vivliostyle.vercel.app/#src=data:,%3Cstyle%3E%20:root%20%7B%20%20%20color:%20rgb(0%25,%2050%25,%200%25);%20%7D%20%3C/style%3E%20%3Cp%3EThis%20text%20should%20be%20green.%3C/p%3E
  ![Screen Shot 2022-10-03 at 23 57 15](https://user-images.githubusercontent.com/3324737/193609704-ee22031e-59f0-4539-ab9e-ceda0aac94e2.png)

```html
<style>
:root {
  --color: rgb(0%, 50%, 0%);
  color: var(--color);
}
</style>
<p>This text should be green.</p>
```

- Before this fix (v2.18.3): https://vivliostyle.github.io/viewer/v2.18.3/#src=data:,%3Cstyle%3E%20:root%20%7B%20%20%20--color:%20rgb(0%25,%2050%25,%200%25);%20%20%20color:%20var(--color);%20%7D%20%3C/style%3E%20%3Cp%3EThis%20text%20should%20be%20green.%3C/p%3E
- After this fix: https://vivliostyle-git-fix-issue1012-vivliostyle.vercel.app/#src=data:,%3Cstyle%3E%20:root%20%7B%20%20%20--color:%20rgb(0%25,%2050%25,%200%25);%20%20%20color:%20var(--color);%20%7D%20%3C/style%3E%20%3Cp%3EThis%20text%20should%20be%20green.%3C/p%3E

**Comment by MurakamiShinyu** [comment]
> In the repro, it seems that the color values are being rendered to RGB instead of maintaining the variable references:

CSS property cascading processing with variable substitution is implemented in Vivliostyle.js/core, and the result property value is set to the element in DOM. For example, when the color property value after variable substitution is `hsl(0, 65%, 65%)`, the element style is set by `element.style.color = "hsl(0, 65%, 65%)"`. In the result, the value of `element.style.color` becomes `rgb(224, 108, 108)` because the computed value of `hsl(0, 65%, 65%)` is serialized as `rgb(224, 108, 108)`. (see https://www.w3.org/TR/css-color-4/#hsl-to-rgb and https://www.w3.org/TR/css-color-4/#serializing-sRGB-values)

**Comment by Enteleform** [comment]
@MurakamiShinyu

Thanks for the quick fix! 🙌


&nbsp;  
> CSS property cascading processing with variable substitution is implemented in Vivliostyle.js/core, and the result property value is set to the element in DOM.

Do you know of any circumstances that might prevent substitution in some cases? I discovered the original issue in an actual project (*during both `vivliostyle build` & `vivliostyle preview`*), where the substitutions were not being made in some cases. In the more minimal repro context, the substitutions were made as you described.


&nbsp;
> the element style is set by `element.style.color =`

I've run into another issue related to this, where the combination of variable hoisting to `element.style` and variables with the same name at differing levels of scope ends up getting broken due to recursive references that are introduced by the altered stylesheet structure. These issues are not present when viewing the page without the Vivliostyle viewer.

### Issue: shared name recursion
(*you can also see here one of the cases where variables are not substituted, @ the first clicked variable*)

https://user-images.githubusercontent.com/10906415/193676483-aeb2a746-3239-4460-b111-c79406b350e0.mp4

### Workaround: unique names

https://user-images.githubusercontent.com/10906415/193676481-cc90c069-3b16-4239-8853-4e0a7813c873.mp4


&nbsp;  
I'll see if I can recreate that in a minimal repro, also have a couple of other things I ran into that I'll try to narrow down.

**Comment by MurakamiShinyu** [comment]
@Enteleform Please open a new issue if you think Vivliostyle.js's CSS variable processing has problem.

The bug I fixed in #1013 (percentages in color values, `rgb()`, `hsl()`, etc.,  at `:root` were changed to vw units) existed since a long time ago (found in https://github.com/sorotokin/adaptive-layout/blob/master/js/adapt/csscasc.js#L448-L449 in [adaptive-layout](https://github.com/sorotokin/adaptive-layout) from which Vivliostyle.js was forked). Thank you for finding it. On the other hand, Vivliostyle.js's CSS variable support is very new (added in [v2.17.0 (2022-07-29)](https://github.com/vivliostyle/vivliostyle.js/blob/master/CHANGELOG.md#2170-2022-07-29)). These should be separate problems.

**Comment by Enteleform** [comment]
@MurakamiShinyu

Was actually wrapping up some repros when you commented, just submitted them:
- https://github.com/vivliostyle/vivliostyle.js/issues/1014
- https://github.com/vivliostyle/vivliostyle.js/issues/1015
